### PR TITLE
Core: Add InMemoryCatalog

### PR DESCRIPTION
**Pull request by nastra**
Having an in-memory catalog might be useful in certain tests, so this is a revival of #4054 with some adjustments.

### CODE REVIEW

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java [hunk 1]
```diff
@@ -736,7 +736,7 @@ public void testRenameTableMissingSourceTable() {
     AssertHelpers.assertThrows(
         "Should reject renaming a table that does not exist",
         NoSuchTableException.class,
-        "Table does not exist",
+        "does not exist",
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:739]: Why was this needed?

File: core/src/test/java/org/apache/iceberg/inmemory/TestInMemoryCatalog.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import org.apache.iceberg.catalog.CatalogTests;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.junit.jupiter.api.BeforeEach;
+
+public class TestInMemoryCatalog extends CatalogTests<InMemoryCatalog> {
+  private InMemoryCatalog catalog;
+
+  @BeforeEach
+  public void before() {
+    catalog = new InMemoryCatalog();
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/inmemory/TestInMemoryCatalog.java:30]: Instance variable assignment should use `this.` to make it clear that it isn't setting a local variable.

File: core/src/test/java/org/apache/iceberg/inmemory/InMemoryCatalog.java [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+
+  @Override
+  protected TableOperations newTableOps(TableIdentifier tableIdentifier) {
+    return new InMemoryTableOperations(io, tableIdentifier);
+  }
+
+  @Override
+  protected String defaultWarehouseLocation(TableIdentifier tableIdentifier) {
+    return SLASH.join(
+        defaultNamespaceLocation(tableIdentifier.namespace()), tableIdentifier.name());
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/inmemory/InMemoryCatalog.java:95]: Did you intend to include the warehouse location? It doesn't matter much since all IO is in memory, but the warehouse location is pulled out of properties and is available.

File: core/src/test/java/org/apache/iceberg/inmemory/InMemoryCatalog.java [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+/**
+ * Catalog implementation that uses in-memory data-structures to store the namespaces and tables.
+ * This class doesn't touch external resources and can be utilized to write unit tests without side
+ * effects. It uses {@link InMemoryFileIO}.
+ */
+public class InMemoryCatalog extends BaseMetastoreCatalog implements SupportsNamespaces, Closeable {
+  private static final Joiner SLASH = Joiner.on("/");
+  private static final Joiner DOT = Joiner.on(".");
+
+  private final ConcurrentMap<Namespace, Map<String, String>> namespaceDb;
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/inmemory/InMemoryCatalog.java:62]: It seems odd to use the suffix "Db" here since it's just a map. I'd probably use a simpler name, like `namespaces` and `tables` for tables.

File: core/src/test/java/org/apache/iceberg/inmemory/InMemoryCatalog.java [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+  public boolean setProperties(Namespace namespace, Map<String, String> properties)
+      throws NoSuchNamespaceException {
+    if (!namespaceExists(namespace)) {
+      throw new NoSuchNamespaceException("Namespace does not exist: %s", namespace);
+    }
+
+    namespaceDb.computeIfPresent(
+        namespace,
+        (k, v) -> {
+          Map<String, String> newProperties = Maps.newHashMap(v);
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/inmemory/InMemoryCatalog.java:218]: Should we use ImmutableMap.Builder just in case?

File: core/src/test/java/org/apache/iceberg/inmemory/InMemoryCatalog.java [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+
+    if (!namespaceExists(toTableIdentifier.namespace())) {
+      throw new NoSuchNamespaceException(
+          "Cannot rename %s to %s because the namespace %s does not exist",
+          fromTableIdentifier, toTableIdentifier, toTableIdentifier.namespace());
+    }
+
+    if (!tableDb.containsKey(fromTableIdentifier)) {
+      throw new NoSuchTableException(
+          "Cannot rename %s to %s because the table %s does not exist",
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/inmemory/InMemoryCatalog.java:154]: We'd normally use `:` instead of "because".

File: core/src/test/java/org/apache/iceberg/inmemory/TestInMemoryCatalog.java (hunk 2 above)
Comment (nastra) [reply core/src/test/java/org/apache/iceberg/inmemory/TestInMemoryCatalog.java:30]: done

File: core/src/test/java/org/apache/iceberg/inmemory/InMemoryCatalog.java (hunk 4 above)
Comment (nastra) [reply core/src/test/java/org/apache/iceberg/inmemory/InMemoryCatalog.java:62]: done

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java (hunk 1 above)
Comment (nastra) [reply core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:739]: I've  updated the error msg so that we don't need that change

File: core/src/test/java/org/apache/iceberg/inmemory/InMemoryCatalog.java (hunk 5 above)
Comment (nastra) [reply core/src/test/java/org/apache/iceberg/inmemory/InMemoryCatalog.java:218]: makes sense, I've changed this to use `ImmutableMap.<String, String>builder().putAll(v).putAll(properties).buildKeepingLast()`

**Comment by nastra** [comment]
@rdblue is there anything else that needs to be addressed or do you think it's good to go?

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @nastra!
